src/rtc.c

```diff
@@ -0,0 +1,518 @@
+
+/**
+ * @defgroup rtc Real-Time Clock Subsystem
+ * @ingroup libdragon
+ * @brief Joybus real-time clock interface.
+ *
+ * The Joybus real-time clock is an accessory that was only included on
+ * one official cartridge that was only available in Japan. The Joybus RTC
+ * is accessed through the serial interface (SI) similar to EEPROM.
+ * A real RTC uses a battery to power a clock that runs even when the N64
+ * is powered-off and maintains the date, time, and day of the week.
+ *
+ * To check if the real-time clock is available, call #joybus_rtc_is_present.
+ * If detected, it is a good idea to ensure the RTC control block is in a
+ * known-good state by calling the #joybus_rtc_run function before reading
+ * the current time with #joybus_rtc_read_time for the first time.
+ *
+ * The simplest method to change the time is by calling #joybus_rtc_set,
+ * which is a high-level convenience function that prepares the RTC and
+ * writes the time with the proper delays.
+ *
+ * In general, since only one game ever used Joybus RTC (and that game was
+ * later re-released on the GameCube in English), RTC support by emulators
+ * and flash carts tends to be incomplete, inaccurate, or non-existent.
+ * Many emulators do not actually implement the RTC write command and will
+ * always respond with the host system's current local time. Some emulators
+ * and flash carts support writing to RTC but will not persist the date/time
+ * after resetting or powering-off.
+ *
+ * The only way to check if writes are actually supported is to write a
+ * time to the RTC and read the time back out. Many emulators that do
+ * support RTC reads will silently ignore RTC writes. You should detect
+ * whether writes are supported using #joybus_rtc_is_writable so that you can
+ * conditionally show the option to change the time if it's supported.
+ *
+ * The RTC sends and receives the date/time in bytes of binary-coded decimal
+ * numbers. This subsystem handles decoding and encoding the date/time into
+ * a more generally-useful struct called #rtc_time_t.
+ *
+ * Special thanks to marshallh and jago85 for their hard work and research
+ * reverse-engineering and documenting the inner-workings of the Joybus RTC.
+ *
+ * @{
+ */
+
+/**
+ * @brief Real-time clock identifier byte.
+ *
+ * A magic value in the RTC status response that indicates the RTC is present.
+ */
+#define JOYBUS_RTC_IDENTIFIER 0x10
+
+/**
+ * @brief Duration (in milliseconds) to wait after writing to RTC.
+ *
+ * The software must wait for the previous RTC write to finish before issuing
+ * another RTC write command. Since there is no status returned by the RTC to
+ * indicate that a write has completed, the software must wait a fixed duration.
+ */
+#define JOYBUS_RTC_WRITE_DELAY 20
+
+/**
+ * @brief The RTC is ready for block 2 to be read.
+ */
+#define JOYBUS_RTC_STATUS_READY 0x00
+/**
+ * @brief The RTC is ready for block 2 to be written.
+ */
+#define JOYBUS_RTC_STATUS_BUSY  0x80
```

Comment:
Comment and variable name don't seem to match

Reply:
Interestingly, as far as I can tell this comment is correct. When the RTC is in "set mode", you have to wait for the busy status before you can write to it. Once the RTC is put back into "run mode" you have to wait for the busy status to clear before continuing.

The name "busy" is consistent with [other documentation I have found](https://github.com/BluRosie/doubutsu-no-mori/blob/default/docs/disassembled%20RTC%20access%20routines.txt)

Reply:
If the bit is really called "busy", maybe its semantic meaning is reversed? I think it's normal to wait for a busy bit to be *off*, or of a ready bit to be *on*. I think calling it "ready" is more clear if we wait for it before writing.

Reply:
I think the most accurate name for this status is actually "STOPPED".

This is consistent with the [documentation of the `0x0004` control register bit as "disable clock"](https://github.com/BluRosie/doubutsu-no-mori/blob/default/docs/Clock%20Access.txt)